src/CacheManager.h

```diff
@@ -84,26 +84,22 @@ void CacheManager<Data>::startPrefetch(DPHolder<Data> dp_holder, CkCallback cb)
 
 template <typename Data>
 void CacheManager<Data>::startParentPrefetch(DPHolder<Data> dp_holder, CkCallback cb) {
-  std::set<Key> request_list;
-  for (Key k : prefetch_set) {
-    for (int i = 0; i < BRANCH_FACTOR; i++) {
-      request_list.insert(k * BRANCH_FACTOR + i);
-    }
-  }
-  request_list.insert(1);
-  std::vector<Key> flat_rl (request_list.size());
-  std::copy(request_list.begin(), request_list.end(), flat_rl.begin());
-  dp_holder.proxy.request(flat_rl.data(), flat_rl.size(), this->thisIndex, cb);
+  std::vector<Key> request_list (prefetch_set.begin(), prefetch_set.end());
+  dp_holder.proxy.request(request_list.data(), request_list.size(), this->thisIndex, cb);
 }
 
 template <typename Data>
 void CacheManager<Data>::prepPrefetch(Node<Data>* node) {
-  // THIS IS IN LOCK
+  // THIS IS IN LOCK. make faster? TODO
   nodewide_data += node->data;
   Key curr_key = node->key;
-  while (curr_key > 1) {
-    curr_key /= BRANCH_FACTOR;
+  auto branch_factor = node->getBranchFactor();
+  while (curr_key >= 1) {
+    curr_key /= branch_factor;
     prefetch_set.insert(curr_key);
+    for (int i = 0; i < branch_factor; i++) {
+      prefetch_set.insert(curr_key * branch_factor + i);
+    }
```

Comment:
Wouldn't this be problematic when `curr_key` is 1 (the root)? After `curr_key /= branch_factor`, `curr_key` would be 0.

Reply:
yeah can you fix it